src/rust/engine/src/nodes.rs

```diff
@@ -1034,47 +1079,79 @@ impl Node for NodeKey {
         stderr: None,
       };
 
-      let started_workunit_id = context
-        .session
-        .workunit_store()
-        .start_workunit(span_id, name, parent_id, metadata);
-      (started_workunit_id, user_facing_name)
+      let started_workunit_id =
+        context
+          .session
+          .workunit_store()
+          .start_workunit(span_id, name, parent_id, metadata.clone());
+      (started_workunit_id, user_facing_name, metadata)
     };
 
     scope_task_workunit_state(Some(workunit_state), async move {
       let context2 = context.clone();
-      let maybe_watch = if let Some(path) = self.fs_subject() {
+      if let Some(path) = self.fs_subject() {
         let abs_path = context.core.build_root.join(path);
-        context
+        let _ = context
           .core
           .watcher
           .watch(abs_path)
           .map_err(|e| Context::mk_error(&format!("{:?}", e)))
-          .await
-      } else {
-        Ok(())
-      };
+          .map_err(|failure| {
+            if let Some(user_facing_name) = user_facing_name {
+              failure.with_pushed_frame(&user_facing_name)
+            } else {
+              failure
+            }
+          })
+          .await?;
```

Comment:
The weird nesting here before was because bailing out early like this will cause the workunit to start but never finish... probably needs to go back how it was.